### ISSUE TITLE
Fuzzy clustering gives weird results when using DataFrame

### ISSUE THREAD

**Issue by helske** [scikit-learn-contrib/hdbscan#161]
In below, I create some fake data which I convert to a Pandas DataFrame. If I the take pretend that I take subset of rows from the data, while actually subsetting the whole data, the membership probabilities from these two cases do not match:

```
from sklearn.datasets import make_blobs
import pandas as pd
import hdbscan

# generate fake data
ids = ["id" + str(s + 1) for s in range(100)]
blobs, labels = make_blobs(random_state=1)
data1 = pd.DataFrame(blobs, index=ids)
data2 = data1.loc[ids, ]
print("DataFrames are equal: " + str(data1.equals(data2)))
clusterer = hdbscan.HDBSCAN(prediction_data=True).fit(data1)
clusterer2 = hdbscan.HDBSCAN(prediction_data=True).fit(data2)
print("\nInput data are equal: " +
    str(np.array_equal(clusterer._raw_data, clusterer2._raw_data)))
print("\nResulting probabilities are equal: " +
    str(np.array_equal(clusterer.probabilities_, clusterer2.probabilities_)))

membership_probabilities = hdbscan.all_points_membership_vectors(clusterer)
membership_probabilities2 = hdbscan.all_points_membership_vectors(clusterer2)

print("\nMembership vectors are not equal: ")
print(membership_probabilities[0])
print(membership_probabilities2[0])
```
Output:

```
DataFrames are equal: True

Input data are equal: True

Resulting probabilities are equal: True

Membership vectors are not equal: 
[0.26227276 0.03622283 0.04050666]
[0.07871163 0.16985801 0.09043261]
```

I am rather new user of Python, so I am not sure if this is more to do with Pandas or with HDBSCAN.

Not sure if related, but I also tested the method `hdbscan.membership_vector` with the same original data, and got completely different results and a warning (I had to use `values` method for the datasets, otherwise I got an error):

```
membership_probabilities_b = hdbscan.membership_vector(clusterer, data1.values)
membership_probabilities2_b = hdbscan.membership_vector(clusterer2, data2.values)
print("\nMembership vectors are not equal: ")
print(membership_probabilities_b[0])
print(membership_probabilities2_b[0])
Membership vectors are not equal: 
[0.25575627 0.04046017 0.04278581]
[0.1628558  0.10183887 0.07430758]
/anaconda3/lib/python3.6/site-packages/hdbscan/prediction.py:482: RuntimeWarning: overflow encountered in exp
  clusterer.prediction_data_.cluster_tree)
```
Using the original `blobs` gives the same result as the previous case with `data1` while still giving a warning:
```
membership_probabilities_c = hdbscan.membership_vector(clusterer, blobs)
membership_probabilities2_c = hdbscan.membership_vector(clusterer2, blobs)
print("\nMembership vectors are equal: ")
print(membership_probabilities_c[0])
print(membership_probabilities2_c[0])
Membership vectors are equal: 
[0.25575627 0.04046017 0.04278581]
[0.25575627 0.04046017 0.04278581]
/anaconda3/lib/python3.6/site-packages/hdbscan/prediction.py:482: RuntimeWarning: overflow encountered in exp
  clusterer.prediction_data_.cluster_tree)
```

**Comment by lmcinnes** [comment]
The membership vector stuff is a little buggy and I haven't ever quite gotten around to tracking down all the errors. It mostly works, which makes finding the issues harder. I guess I would class it as "use with caution" for now.

**Comment by m-dz** [comment]
@lmcinnes , it's not membership vector, it's pandas loc/iloc stuff...

With the simple change below everything works...

```
ids = ["id" + str(s + 1) for s in range(100)]
iids = range(100)
blobs, labels = make_blobs(random_state=1)
data1 = pd.DataFrame(blobs, index=ids)
data2 = data1.iloc[iids, ]
```

On the other hand, with the `loc` and `ids` version everything seems to be fine up to the `min_dist_to_exemplar` function, so there might be some issues with memory allocation or casting there.

**Comment by m-dz** [comment]
To follow up, after changing `min_dist_to_exemplar` to `cpdef` here:
https://github.com/scikit-learn-contrib/hdbscan/blob/4774b2b143b811800e3eda4d3eaa3d426939fee3/hdbscan/_prediction_utils.pyx#L25
to be able to import it directly, the following code sheds some light on the problem and hopefully will help to solve the problem (I give up, all this memory magic is too much):

```
from hdbscan._prediction_utils import min_dist_to_exemplar
from sklearn.datasets import make_blobs

import pandas as pd
import numpy as np
import hdbscan


ids = ["id" + str(s + 1) for s in range(100)]
iids = range(100)
blobs, labels = make_blobs(random_state=1)
data1 = pd.DataFrame(blobs, index=ids)
data2 = data1.loc[ids, ]
data3 = data1.iloc[iids, ]
clusterer = hdbscan.HDBSCAN(prediction_data=True).fit(data1)
clusterer2 = hdbscan.HDBSCAN(prediction_data=True).fit(data2)
clusterer3 = hdbscan.HDBSCAN(prediction_data=True).fit(data3)

print(np.array_equal(
    clusterer.prediction_data_.raw_data[0],
    clusterer2.prediction_data_.raw_data[0]
))

print(min_dist_to_exemplar(
    clusterer.prediction_data_.raw_data[0],
    clusterer.prediction_data_.exemplars[0],
    clusterer.prediction_data_.dist_metric
))

print(min_dist_to_exemplar(
    clusterer2.prediction_data_.raw_data[0],
    clusterer.prediction_data_.exemplars[0],
    clusterer.prediction_data_.dist_metric
))

print(min_dist_to_exemplar(
    clusterer3.prediction_data_.raw_data[0],
    clusterer.prediction_data_.exemplars[0],
    clusterer.prediction_data_.dist_metric
))
```

**Comment by lmcinnes** [comment]
Thanks for this. I'll try to take a closer look when I get some time to dig into it properly. The help is greatly appreciated.

**Comment by helske** [comment]
Yes I noticed that with `iloc` everything seems to work (except I still get overflow warning). But the issue is not just with `loc` if I define the second DataFrame as `data2 = pd.DataFrame(data1, copy=True)` I get wrong results again.

**Comment by helske** [comment]
In the above example by @m-dz, changing the second case to `copy.deepcopy(clusterer2.prediction_data_.raw_data[0])`everything seems to work, or if I just assign `clusterer2.prediction_data_.raw_data = clusterer.prediction_data_.raw_data`. After digging deeper I ended up to the same conclusion as above that the issue is in `min_dist_to_exemplar` function:

```
cdef np.float64_t min_dist_to_exemplar(
                        np.ndarray[np.float64_t, ndim=1] point,
                        np.ndarray[np.float64_t, ndim=2] cluster_exemplars,
                        DistanceMetric dist_metric):

    cdef np.intp_t i
    cdef np.float64_t result = DBL_MAX
    cdef np.float64_t distance
    cdef np.float64_t *point_ptr = (<np.float64_t *> point.data)
    cdef np.float64_t[:, ::1] exemplars_view = \
        (<np.float64_t [:cluster_exemplars.shape[0], :cluster_exemplars.shape[1]:1]> (<np.float64_t *> cluster_exemplars.data))
    cdef np.float64_t *exemplars_ptr = \
        (<np.float64_t *> &exemplars_view[0, 0])
    cdef np.intp_t num_features = point.shape[0]

    for i in range(cluster_exemplars.shape[0]):
        distance = dist_metric.dist(point_ptr,
                                    &exemplars_ptr[num_features * i],
                                    num_features)
        if distance < result:
            result = distance

    return result
```
All the outputs from `dist_metric.dist`differ between the two cases, and the second and third arguments seem to be identical
Unfortunately I don't understand the Cython stuff well enough to pinpoint the problem, but if I fill `cluster_exemplars`inside of `min_dist_to_exemplar` to say 1, then I get identical results in all cases.

**Comment by helske** [comment]
In `loc` case, computing the euclidean distance "by hand" as `np.sum(np.square(point-cluster_exemplars[num_features * i]))` gives different results than the above code in case the number of features is > 1. This is weird as copying the exemplars from `clusterer` does not help, while copying the raw data does.

I am out of ideas. I would need the fuzzy clustering in my research, but for now I'll tircumvent the issue by working only with numpy arrays, where I don't see obvious errors.

**Comment by dpatschke** [comment]
Forgive me if this is not the correct place to ask this but this seems to be a "membership_vector" issue and I couldn't find this information after searching for a little while and didn't really want to open up a new issue.

I'm interested in getting the outlier probabilities on new data points (a test set). I thought that perhaps I could derive the outlier values by applying the 'membership_vector' function, summing the rows of the resulting matrix, and subtracting that value from '1' for each row. However, when I compare this difference value to the 'outlier_scores_' value returned as part of the cluster object, it doesn't always match up.

@lmcinnes Is this what is meant by 'use with caution'? 

Examples:
From cluster object:

|    |   cluster |     prob |    outlier |
|---:|----------:|---------:|-----------:|
|  0 |         2 | 1        | 0.00704651 |
|  1 |         2 | 1        | 0.0650792  |
|  2 |         2 | 1        | 0.0455353  |
|  3 |         2 | 0.965305 | 0.0346955  |
|  4 |         2 | 1        | 0.00650369 |
|  5 |         2 | 1        | 0          |
|  6 |         2 | 1        | 0.034998   |
|  7 |         2 | 1        | 0          |
|  8 |         1 | 0.893451 | 0.106549   |
|  9 |        -1 | 0        | 0.307274   |

From approximate_predict:

|    |   cluster_num |   cluster_prob |
|---:|--------------:|---------------:|
|  0 |             2 |       1        |
|  1 |             2 |       1        |
|  2 |             2 |       1        |
|  3 |             2 |       0.965305 |
|  4 |             2 |       1        |
|  5 |             2 |       1        |
|  6 |             2 |       1        |
|  7 |             2 |       1        |
|  8 |             1 |       0.893451 |
|  9 |            -1 |       0        |

From membership_vector:

|    |   cluster0_prob |   cluster1_prob |   cluster2_prob |   outlier_prob |
|---:|----------------:|----------------:|----------------:|---------------:|
|  0 |    2.18923e-120 |    1.97513e-119 |     1           |    1.53417e-09 |
|  1 |    1.94868e-11  |    4.8861e-11   |     1           |    1.44325e-09 |
|  2 |    3.42036e-16  |    3.38147e-15  |     1           |    1.59603e-09 |
|  3 |    4.02472e-21  |    3.80827e-19  |     0.965305    |    0.0346955   |
|  4 |    1.29005e-131 |    3.40929e-131 |     1           |    1.35105e-09 |
|  5 |    0.402237     |    0.597763     |     0           |    1.15902e-09 |
|  6 |    1.64402e-22  |    6.60667e-22  |     1           |    1.489e-09   |
|  7 |    0            |    0            |     1           |    1.06918e-09 |
|  8 |    2.068e-05    |    0.893278     |     0.000151819 |    0.106549    |
|  9 |    0.182153     |    0.169257     |     0.188788    |    0.459803    |

Thanks,
David

**Comment by lmcinnes** [comment]
By "use with caution" I think I mean "go ahead and try it, but do be sure to validate any results, and if thing look weird be prepared to toss the results". The algorithm is generally sound, but the code needs an overhaul. More importantly I've been working on an alternative to hdbscan that would make all of this far more intuitive and easy, so that leaves me loathe to work on code that may become obsolete soon (the membership vector aspects -- hdbscan itself is fine).

**Comment by helske** [comment]
Hi @lmcinnes I'd like to discuss a bit about this and related issues, can I send you an email? I didn't find your address anywhere.

**Comment by lmcinnes** [comment]
You can reach me at leland.mcinnes@gmail.com if you need to.
